enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/document/DocumentProcessingService.java

```diff
@@ -119,83 +119,139 @@ public List<DocumentVerificationEntity> submitDocuments(
             DocumentSubmitRequest request,
             OwnerId ownerId) throws DocumentSubmitException {
 
-        List<Document> documents = getDocuments(ownerId, request);
-
-        List<DocumentVerificationEntity> docVerifications = new ArrayList<>();
-        List<DocumentResultEntity> docResults = new ArrayList<>();
+        checkDocumentResubmit(ownerId, request);
+        final List<Document> documents = getDocuments(ownerId, request);
+        final var documentsByType = request.getDocuments().stream()
+                .collect(groupingBy(DocumentSubmitRequest.DocumentMetadata::getType));
 
-        List<DocumentSubmitRequest.DocumentMetadata> docsMetadata = request.getDocuments();
-        for (DocumentSubmitRequest.DocumentMetadata docMetadata : docsMetadata) {
-            DocumentVerificationEntity docVerification = createDocumentVerification(ownerId, idVerification, docMetadata);
-            docVerification.setIdentityVerification(idVerification);
-            docVerifications.add(docVerification);
+        return documentsByType.values().stream()
+                .map(v -> submitDocument(v, documents, idVerification, ownerId))
```

Comment:
❓ `.map` has a side effect. Isn't it designed as a data transformer only?

Reply:
You're right, using stateful function or function with a side effect as a mapping function in stream is not good practice.